overlord/snapstate/snapstate_update_test.go

```diff
@@ -9642,3 +9625,92 @@ func (s *snapmgrTestSuite) TestUnlinkMonitorSnapOnHardCheckFailure(c *C) {
 	c.Assert(monitored, DeepEquals, map[string]bool{"some-snap": true})
 	close(monitorSignal)
 }
+
+func (s *snapmgrTestSuite) TestDeletedMonitoredMapIsCorrectlyDeleted(c *C) {
+	s.state.Lock()
+	si := &snap.SideInfo{
+		RealName: "foo",
+		SnapID:   "foo-id",
+		Revision: snap.R(1),
+	}
+	snaptest.MockSnap(c, `name: foo`, si)
+	snapstate.Set(s.state, "foo", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{si},
+		Current:  si.Revision,
+	})
+	snapsup := &snapstate.SnapSetup{
+		SideInfo: &snap.SideInfo{
+			RealName: "foo",
+			Revision: snap.R(2),
+		},
+		Flags: snapstate.Flags{IsAutoRefresh: true},
+	}
+
+	restore := snapstate.MockRefreshAppsCheck(func(info *snap.Info) error {
+		return snapstate.NewBusySnapError(info, []int{123}, nil, nil)
+
+	})
+	defer restore()
+
+	var monitorSignal chan<- string
+	restore = snapstate.MockCgroupMonitorSnapEnded(func(name string, done chan<- string) error {
+		c.Check(name, Equals, "foo")
+		monitorSignal = done
+		return nil
+	})
+	defer restore()
+
+	preDlChg := s.state.NewChange("pre-download", "pre-download change")
+	preDlTask := s.state.NewTask("pre-download-snap", "pre-download task")
+
+	preDlTask.Set("snap-setup", snapsup)
+	preDlTask.Set("refresh-info", &userclient.PendingSnapRefreshInfo{InstanceName: "foo"})
+	preDlChg.AddTask(preDlTask)
+
+	s.settle(c)
+	c.Assert(preDlChg.Status(), Equals, state.DoneStatus)
+
+	// unblock the monitoring routine which will delete the "monitored-snaps" map
+	s.state.Unlock()
+	monitorSignal <- "foo"
+	waitForMonitoringEnd(s.state, c)
+
+	s.state.Lock()
+	c.Assert(s.state.Cached("monitored-snaps"), IsNil)
+
+	// start a 2nd task that checks the state for the map of monitored snap
+	preDlChg = s.state.NewChange("pre-download", "pre-download change")
+	preDlTask = s.state.NewTask("pre-download-snap", "pre-download task")
+	preDlTask.Set("snap-setup", snapsup)
+	preDlTask.Set("refresh-info", &userclient.PendingSnapRefreshInfo{InstanceName: "foo"})
+	preDlChg.AddTask(preDlTask)
+
+	s.settle(c)
+	c.Assert(preDlChg.Status(), Equals, state.DoneStatus)
+
+	s.state.Unlock()
+	monitorSignal <- "foo"
+	waitForMonitoringEnd(s.state, c)
+}
+
+func waitForMonitoringEnd(st *state.State, c *C) {
+	for i := 0; i < 5; i++ {
+		select {
+		case <-time.After(time.Second):
+			st.Lock()
+			finished := st.Cached("auto-refresh-continue-attempt")
+			if finished == nil || finished.(int) != 1 {
+				st.Unlock()
+				continue
+			}
+
+			c.Assert(st.Cached("monitored-snaps"), IsNil)
+			c.Check(st.Cached("auto-refresh-continue-attempt"), Equals, 1)
```

Comment:
Personal preference nitpick: I think it would be more intuitive if these two checks were in the main test instead of hidden in the helper function `waitForMonitoringEnd` which doesn't imply that these checks are made.

Reply:
Fwiw, I like this suggestion